### PR TITLE
Feature/mobilenetv2ssd support

### PR DESCRIPTION
**Pull request by jsalas98**

### CODE REVIEW

**Review by rrcarlosrodriguez — COMMENTED** [review]
Please make sure we are not leaving any memory behind. A quick valgrind memory test with a basic pipeline should be enough

File: ext/r2inference/gstmobilenetv2ssd.c [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+  GST_OBJECT_UNLOCK (mobilenetv2ssd);
+
+  for (i_box = 0; i_box < num_boxes; i_box++) {
+    /* Here prediction has the 4 concatenated tensors in the order
+       [locations, labels, probabilities, num_boxes], so we compute the indices 
+       to access each one of them */
+    i_location = i_box * LOCATION_PARAMS;
+    i_label = i_box + (num_boxes * LOCATION_PARAMS);
+    i_prob = i_label + num_boxes;
+    prob = ((gfloat *) (prediction))[i_prob];
```
Comment (fasm22) [inline ext/r2inference/gstmobilenetv2ssd.c:215]: Is it possible to cast against gfloat only once? You use this pointer in the rest of this code section multiple times

File: ext/r2inference/gstmobilenetv2ssd.c [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+
+    if (prob > prob_thresh) {
+      BBox result = { 0 };
+      /* We create an array of probabilities of the total classes size to have
+         compatibility with the other classification models, but this model only
+         outputs 1 label and 1 probability per bounding box */
+      gdouble *cur_prob = g_malloc0 (TOTAL_CLASSES * sizeof (gdouble));
```
Comment (fasm22) [inline ext/r2inference/gstmobilenetv2ssd.c:222]: When are we releasing this memory?

**Review by fasm22 — CHANGES_REQUESTED** [review]

File: ext/r2inference/gstmobilenetv2ssd.c (hunk 2 above)
Comment (jsalas98) [reply ext/r2inference/gstmobilenetv2ssd.c:222]: It is released in the postprocess function after it is used, see line 311.

**Review by fasm22 — APPROVED** [review]

File: ext/r2inference/gstmobilenetv2ssd.c [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+        gst_create_prediction_from_box (vi, &boxes[i], labels_list, num_labels,
+        probabilities[i]);
+    gst_inference_prediction_append (imeta->prediction, pred);
+    g_free (probabilities[i]);
+  }
+
+  *valid_prediction = (valid_boxes > 0) ? TRUE : FALSE;
+
+  if (0 == valid_boxes) {
+    goto out;
```
Comment (michaelgruner) [inline ext/r2inference/gstmobilenetv2ssd.c:317]: We need to free stuff here.

File: ext/r2inference/gstmobilenetv2ssd.c [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+     is of shape [N * 4]
+     1: [N] tensor with the number of labels for the N bounding boxes
+     2: [N] tensor with the probabilities of those N labels
+     3: [1] tensor with the number of detected boxes 
+     They are all concatenated here in a 1D array in row-major order.
+   */
+  total_boxes = pred[predsize / sizeof (gfloat) - 1];
+
+  boxes = g_malloc (total_boxes * sizeof (BBox));
+  probabilities = g_malloc (total_boxes * sizeof (gdouble));
```
Comment (michaelgruner) [inline ext/r2inference/gstmobilenetv2ssd.c:291]: What happens if total boxes is 0?

File: ext/r2inference/gstmobilenetv2ssd.c [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+  g_return_val_if_fail (outframe, FALSE);
+
+  GST_LOG_OBJECT (vi, "Preprocess");
+
+  return gst_normalize (inframe, outframe, MEAN, STD, MODEL_CHANNELS);
+}
+
+static gint
+gst_mobilenetv2ssd_get_boxes_from_prediction (GstMobilenetv2ssd *
+    mobilenetv2ssd, gfloat * prediction, gint num_boxes, gint img_width,
```
Comment (michaelgruner) [inline ext/r2inference/gstmobilenetv2ssd.c:188]: I believe our standard is to use `self`

File: ext/r2inference/gstmobilenetv2ssd.c [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+      boxes[cur_box] = result;
+      probabilities[cur_box] = cur_prob;
+      cur_box++;
+    }
+  }
+
+  gst_remove_duplicated_boxes (iou_thresh, boxes, &cur_box);
+  /* Resize the array to match the actual number of valid boxes */
+  boxes = g_realloc (boxes, cur_box * sizeof (BBox));
+  probabilities = g_realloc (probabilities, cur_box * sizeof (gdouble));
```
Comment (michaelgruner) [inline ext/r2inference/gstmobilenetv2ssd.c:246]: No need really. Just unnecessary overhead in my opinion.

File: ext/r2inference/gstmobilenetv2ssd.c [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+gst_mobilenetv2ssd_set_property (GObject * object, guint property_id,
+    const GValue * value, GParamSpec * pspec)
+{
+  GstMobilenetv2ssd *mobilenetv2ssd = GST_MOBILENETV2SSD (object);
+
+  GST_DEBUG_OBJECT (mobilenetv2ssd, "set_property");
+
+  switch (property_id) {
+    case PROP_PROB_THRESH:
+      mobilenetv2ssd->prob_thresh = g_value_get_double (value);
```
Comment (michaelgruner) [inline ext/r2inference/gstmobilenetv2ssd.c:341]: Serialize accesses

File: gst-libs/gst/r2inference/gstinferencepostprocess.c [hunk 8]
```diff
@@ -109,7 +107,7 @@ gst_delete_box (BBox * boxes, gint * num_boxes, gint index)
   }
 }
 
-static void
+void
 gst_remove_duplicated_boxes (gfloat iou_thresh, BBox * boxes, gint * num_boxes)
```
Comment (michaelgruner) [inline gst-libs/gst/r2inference/gstinferencepostprocess.c:111]: Always use double unless you have a special reason not to.

**Review by michaelgruner — CHANGES_REQUESTED** [review]

File: ext/r2inference/gstmobilenetv2ssd.c (hunk 5 above)
Comment (jsalas98) [reply ext/r2inference/gstmobilenetv2ssd.c:188]: I checked the other elements and they do not use the self standard.

File: ext/r2inference/gstmobilenetv2ssd.c [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+
+  g_return_val_if_fail (vi, FALSE);
+  g_return_val_if_fail (prediction, FALSE);
+  g_return_val_if_fail (meta_model, FALSE);
+  g_return_val_if_fail (info_model, FALSE);
+  g_return_val_if_fail (valid_prediction, FALSE);
+
+  GST_LOG_OBJECT (vi, "Postprocess");
+
+  pred = (gfloat *) prediction;
```
Comment (michaelgruner) [inline ext/r2inference/gstmobilenetv2ssd.c:264]: Do we really need to drop the `const`?

File: ext/r2inference/gstmobilenetv2ssd.c [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
+      total_boxes);
+
+  if (0 == total_boxes) {
+    goto out;
+  }
+
+  imeta = (GstInferenceMeta *) meta_model[1];
+
+  boxes = g_malloc (total_boxes * sizeof (BBox));
+  probabilities = g_malloc (total_boxes * sizeof (gdouble));
```
Comment (michaelgruner) [inline ext/r2inference/gstmobilenetv2ssd.c:287]: This is wrong, probabilities is a double pointer. It should be `total_boxes * sizeof (gdouble *)`

File: ext/r2inference/gstmobilenetv2ssd.c [hunk 11]
```diff
@@ -0,0 +1,377 @@
...
+    i_label = i_box + (num_boxes * LOCATION_PARAMS);
+    i_prob = i_label + num_boxes;
+    prob = prediction[i_prob];
+
+    if (prob > prob_thresh) {
+      BBox result = { 0 };
+      /* We create an array of probabilities of the total classes size to have
+         compatibility with the other classification models, but this model only
+         outputs 1 label and 1 probability per bounding box */
+      gdouble *cur_prob = g_malloc0 (TOTAL_CLASSES * sizeof (gdouble));
```
Comment (michaelgruner) [inline ext/r2inference/gstmobilenetv2ssd.c:214]: This size is something we know from outside. When dealing with allocations, keep symmetry. Try not to alloc here and free outside. It's better to alloc and free outside, and fill here.

**Review by michaelgruner — CHANGES_REQUESTED** [review]

**Review by michaelgruner — APPROVED** [review]

**Pull request merged**